subprojects/base-services/src/main/java/org/gradle/internal/hash/Hashing.java

```diff
@@ -196,4 +278,91 @@ public void putHash(HashCode hashCode) {
             putBytes(hashCode.getBytes());
         }
     }
+
+    @VisibleForTesting
+    static class DefaultHasher implements Hasher {
+        private final PrimitiveHasher hasher;
+        private boolean valid = true;
+
+        public DefaultHasher(PrimitiveHasher unsafeHasher) {
+            this.hasher = unsafeHasher;
+        }
+
+        @Override
+        public void putByte(byte value) {
+            hasher.putInt(1);
+            hasher.putByte(value);
+        }
+
+        @Override
+        public void putBytes(byte[] bytes) {
+            hasher.putInt(bytes.length);
+            hasher.putBytes(bytes);
+        }
+
+        @Override
+        public void putBytes(byte[] bytes, int off, int len) {
+            hasher.putInt(len);
+            hasher.putBytes(bytes, off, len);
+        }
+
+        @Override
+        public void putHash(HashCode hashCode) {
+            hasher.putInt(hashCode.length());
+            hasher.putHash(hashCode);
+        }
+
+        @Override
+        public void putInt(int value) {
+            hasher.putInt(4);
+            hasher.putInt(value);
+        }
+
+        @Override
+        public void putLong(long value) {
+            hasher.putInt(8);
+            hasher.putLong(value);
+        }
+
+        @Override
+        public void putDouble(double value) {
+            hasher.putInt(8);
+            hasher.putDouble(value);
+        }
+
+        @Override
+        public void putBoolean(boolean value) {
+            hasher.putInt(1);
+            hasher.putBoolean(value);
+        }
+
+        @Override
+        public void putString(CharSequence value) {
+            hasher.putInt(value.length());
+            hasher.putString(value);
+        }
+
+        @Override
+        public void putNull() {
+            this.putInt(0);
+        }
+
+        @Override
+        public void markAsInvalid() {
+            this.valid = false;
+        }
+
+        @Override
+        public boolean isValid() {
+            return valid;
+        }
+
+        @Override
+        public HashCode hash() {
+            if (!valid) {
+                throw new IllegalStateException("Build cache hash is not valid");
```

Comment:
I guess the exception should be `Hash is not valid`.

Reply:
Yep.